### ISSUE TITLE
stan-to-C++ flag --O1 breaks matrix multiplication

### ISSUE THREAD

**Issue by rmcminds** [stan-dev/cmdstan#1077]
A model that compiles and runs fine in 2.29.0-rc1 with default settings breaks when --O1 is added to the stanc3 compilation. Error "Exception: fma: x ((45, 12)12, 1266) must match in size", referring to a line that contains only matrix multiplication 

Apologies; I don't have a fully reproducible example at the moment, but the apparently relevant code is of the following structure:

data {
  matrix[N1,N2] m1;
  matrix[N3,N4] m4;
  array[N2] int idx;
  array[N3] int idy;
}
parameters {
  matrix[N2,N3] m2;
  matrix[N5,N6] m3;
}
model {
  matrix[N1,N4] m5 = m1 * (m2 .* m3[idx,idy]) + m4
}

#### Current Version:
v2.29.0

**Comment by WardBrian** [comment]
@rok-cesnovar Seems like strange behavior in `stan::math::fma`? I can confirm that Stanc's partial evaluator will produce an fma call, the model line ends up being something like

```C++
        stan::model::assign(m5,
          stan::math::fma(m1,
            stan::math::elt_multiply(m2,
              stan::model::rvalue(m3, "m3",
                stan::model::index_multi(idx), stan::model::index_multi(idy))),
            m4), "assigning variable m5");
```

**Comment by rok-cesnovar** [comment]
This exact model compiles for me :/

**Comment by rok-cesnovar** [comment]
I am not getting the fma in the generated C++ though with 

`bin/stanc --O1 --o=examples/test.hpp examples/test.stan`

**Comment by WardBrian** [comment]
I am pretty sure if you don’t use m5 it gets removed by the optimized. I added target += sum(m5) to force it to exist

**Comment by rmcminds** [comment]
I'll work on a minimal example. Sorry; I had hoped the above would replicate it.

**Comment by rok-cesnovar** [comment]
It does now Ryan. I added Brian's suggested line to the model and that made it break. Updated your post as well.

**Comment by rmcminds** [comment]
ah, great! thanks

**Comment by rok-cesnovar** [comment]
Yeah, I think Stan Math will be the right one. I pointed Ryan here because I had no idea what was up.
